educe/rst_dt/learning/features.py

```diff
@@ -2,390 +2,243 @@
-             MagicKey.discrete_fn(ptb_pos_tag_first_pairs)]
-        super(PairSubgroup_Tuple, self).__init__(desc, keys)
-
-    def fill(self, current, edu1, edu2, target=None):
-        vec = self if target is None else target
-        for key in self.keys:
-            vec[key.name] = key.function(current, self.sf_cache, edu1, edu2)
-
-
-class PairSubgroup_Basket(PairSubgroup):
-    """
-    Sparse features
-    """
-    _features =\
-        [MagicKey.basket_fn(ptb_pos_tags_in_first)]
-
-    def __init__(self):
-        desc = self.__doc__.strip()
-        super(PairSubgroup_Basket, self).__init__(desc, self._features)
-
-
-# export groups
-class SingleEduKeys(BaseSingleEduKeys):
-    """Single EDU features"""
-
-    def __init__(self):
-        groups = [
-            SingleEduSubgroup_Meta(),
-            SingleEduSubgroup_Text(),
-            SingleEduSubgroup_Ptb()
-        ]
-        super(SingleEduKeys, self).__init__(groups)
-
-
-class PairKeys(BasePairKeys):
-    """Features on a pair of EDUs"""
-
-    def __init__(self, sf_cache=None):
-        groups = [
-            PairSubGroup_Core(),
-            PairSubgroup_Gap(),
-            PairSubgroup_Tuple(sf_cache),
-            PairSubgroup_Basket()
-        ]
-        super(PairKeys, self).__init__(groups, sf_cache)
-
-    def init_single_features(self):
-        """Init features defined on single EDUs"""
-        return SingleEduKeys()
+PAIR_GAP = [
+    ('num_edus_between', Substance.CONTINUOUS),
+    ('same_paragraph', Substance.DISCRETE),
+    ('same_bad_sentence', Substance.DISCRETE),
+    ('same_ptb_sentence', Substance.DISCRETE)
+]
+
+
+def extract_pair_gap(edu_info1, edu_info2):
+    """Document tuple features"""
+    edu_num1 = edu_info1['edu'].num
+    edu_num2 = edu_info2['edu'].num
+
+    edu_num_diff = abs(edu_num2 - edu_num1) - 1
+    yield ('num_edus_between', edu_num_diff)
+
+    try:
+        para_id1 = edu_info1['para_idx']
+        para_id2 = edu_info2['para_idx']
+    except KeyError:
+        pass
+    else:
+        same_para = (para_id1 is not None and
+                     para_id2 is not None and
+                     para_id1 == para_id2)
+        yield ('same_paragraph', same_para)
+
+    # same sentence, raw segmentation
+    raw_sent_id1 = edu_info1['raw_sent_idx']
+    raw_sent_id2 = edu_info2['raw_sent_idx']
+    same_bad_sent = (raw_sent_id1 is not None and
+                     raw_sent_id2 is not None and
+                     raw_sent_id1 == raw_sent_id2)
+    yield ('same_bad_sentence', same_bad_sent)
+
+    # same sentence, ptb segmentation
+    # FIXME fix behaviour for sentences with no gold segmentation,
+    # as then edu2sent = edu2raw_sent
+    sent_id1 = edu_info1['sent_idx']
+    sent_id2 = edu_info2['sent_idx']
+    same_ptb_sent = (sent_id1 is not None and
+                     sent_id2 is not None and
+                     sent_id1 == sent_id2)
+    yield ('same_ptb_sentence', same_ptb_sent)
+
```

Comment:
The yield(key, value) pattern is very interesting; hadn't considered that possibility, and it does look like it leads to some very nice simplifications in the code structure for which first, apologies, and second thanks.

It's slightly unfortunate that there's a bit of scatter/duplication in defining the types of the features and the keys (the old KeyGroup code also suffers from this, but tries to minimise it by keeping it in declaration time, and by using the function name as feature name, where possible).

Hmm.  But this looks cleaner overall.  I just wonder if we can somehow avoid the duplication